src/kernel/initrd.zig

```diff
@@ -0,0 +1,673 @@
+pub const RamdiskFS = struct {
+    /// The ramdisk header that stores pointers to the raw ramdisk for the name and file content.
+    /// As the ramdisk is read only, these can be const pointers.
+    const RamdiskHeader = struct {
+        /// The name of the file
+        name: []const u8,
+
+        /// The content of the file
+        content: []const u8,
+    };
+
+    /// The error set for the ramdisk file system.
+    const Error = error{ /// The error for an invalid raw ramdisk when parsing.
+    InvalidRamDisk };
+
+    const Self = @This();
+
+    /// A mapping of opened files so can easily retrieved opened files for reading.
+    opened_files: AutoHashMap(*const vfs.Node, *RamdiskHeader),
+
+    /// The underlying file system
+    fs: *vfs.FileSystem,
+
+    /// The allocator used for allocating memory for opening and reading.
+    allocator: *Allocator,
+
+    /// The list of files in the ram disk. These will be pointers into the raw ramdisk to save on
+    /// allocations.
+    files: []RamdiskHeader,
+
+    /// The root node for the ramdisk file system. This is just a root directory as there is not
+    /// subdirectories.
+    root_node: *vfs.Node,
+
+    /// See vfs.FileSystem.instance
+    instance: usize,
+
+    /// See vfs.FileSystem.getRootNode
+    fn getRootNode(fs: *const vfs.FileSystem) *const vfs.DirNode {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        return &self.root_node.Dir;
+    }
+
+    /// See vfs.FileSystem.close
+    fn close(fs: *const vfs.FileSystem, node: *const vfs.FileNode) void {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        const cast_node = @ptrCast(*const vfs.Node, node);
+        // As close can't error, if provided with a invalid Node that isn't opened or try to close
+        // the same file twice, will just do nothing.
+        if (self.opened_files.contains(cast_node)) {
+            _ = self.opened_files.remove(cast_node);
+            self.allocator.destroy(node);
+        }
+    }
+
+    /// See vfs.FileSystem.read
+    fn read(fs: *const vfs.FileSystem, node: *const vfs.FileNode, len: usize) (Allocator.Error || vfs.Error)![]u8 {
+        var self = @fieldParentPtr(RamdiskFS, "instance", fs.instance);
+        const cast_node = @ptrCast(*const vfs.Node, node);
+        const file_header = self.opened_files.get(cast_node) orelse return vfs.Error.NotOpened;
+        const length = std.math.min(len, file_header.content.len);
+        const buff = try self.allocator.alloc(u8, length);
+        std.mem.copy(u8, buff, file_header.content[0..length]);
+        return buff;
+    }
+
+    /// See vfs.FileSystem.write
+    fn write(fs: *const vfs.FileSystem, node: *const vfs.FileNode, bytes: []const u8) (Allocator.Error || vfs.Error)!void {}
+
+    /// See vfs.FileSystem.open
+    fn open(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8, flags: vfs.OpenFlags) (Allocator.Error || vfs.Error)!*vfs.Node {
```

Comment:
It isn't used no, I wasn't sure how to use it anyway as yes, there are no directories and from your example, it is used for searching, but if there is a use for it, I would love to hear

Reply:
Well the directory node is the directory under which the file (`name` is a file name and extension, not a path) is created/opened, so it can be ignored if you don't plan for the ramdisk to support directories at the moment.

Reply:
I see, the initrd will probably not have any directories, just to make it easier. In that case, shall I rename RamdiskFS to InitrdFS.

Reply:
I'm cool with either